### ISSUE TITLE
Question mark symbols instead of letters when recieving some mails (umlauts Å Ä and Ö)

### ISSUE THREAD

**Issue by ysunil702** [sendgrid/sendgrid-csharp#956]
#### Issue Summary

problem related to the inbound parse webhook functionality offered by SendGrid.
However,  After a lot of searching I found out that if the mail is sent from e.g. hotmail or Outlook (or the like), Å Ä Ö is replaced by question marks.

* sendgrid-csharp Version: 9.5.0
* Platform (ASP.NET)
* .NET Version: 4.5.2

**Comment by thinkingserious** [comment]
Hello @ysunil702,

Thanks for taking the time to report this to us!

I'd like to gather a bit more information to help determine what the issue is:

1. Can you share the code sample you are using to parse the webhook?
2. Were there cases where Å Ä Ö was not replaced by question marks?

Thanks!

With best regards,

Elmer

**Comment by ysunil702** [comment]
Answer 1- 

 	    string contentHtml = null;
            root = HttpContext.Current.Server.MapPath("~/App_Data");
            provider = new MultipartFormDataStreamProvider(root);

            var charsets = provider.FormData.GetValues("charsets").FirstOrDefault();

            if (charsets != null)
            {
                if (charsets.Contains("html"))
                {
                   
                    contentHtml = provider.FormData.GetValues("html").FirstOrDefault();
                }
                else if (charsets.Contains("text"))
                {                       
                    if (provider.FormData.GetValues("text").FirstOrDefault() != null)
                    {
                        contentHtml = provider.FormData.GetValues("text").FirstOrDefault();
     
                        contentHtml = contentHtml.ToHtml();
                    }
                }
            }

Answer 2- When i am sending email from gmail and get the webhook in that cases Å Ä Ö was not replaced by question marks.

**Comment by saurabhabh** [comment]
Do we have an update on this ??

**Comment by thinkingserious** [comment]
Thank you for clarifying @ysunil702!

This does not look to be an issue with this helper library. I would suggest you reach out to our support team at https://support.sendgrid.com.

@saurabhabh,

I suggest that you also reach out to support for help with this issue.

Thanks!

**Comment by Jericho** [comment]
@thinkingserious @childish-sambino You are technically correct that this is not an issue with the client library but that's simply because the library does not provide any way to parse the webhooks received from SendGrid. Also, in my experience your support team will not offer assistance with C# code so instructing your customers to contact support is not helpful.

@ysunil702 @saurabhabh I have experienced the exact same problem you are describing and here is what I found:

Email systems (such as Hotmail for example) can encode emails with a different 'code page' than the usual UTF8 if it contains umlauts and certain other characters. In fact, it's entirely possible that they may encode certain parts of the email with different code pages. For instance: `from`, `to`, `subject`, and `text` are some of the parts that could be encoded differently and it is the responsibility of the receiving mail client to properly decode each section. We normally don't realize that this encoding/decoding is taking place because the software we use to send and receive emails (such as Outlook, Thunderbird, iOS Mail, etc.) automatically encode/decode each part using the appropriate code page. In the case of an "inbound parse webhook", our C# code is responsible for properly decoding each section. Typically it's not a problem because UTF8 is used 99.9% of the time, but it's obviously a problem (as you experienced) when a different code page was used to encode certain parts of the email.

The payload of the webhook that you receive from SendGrid contains an array called `charsets` that tells you which code page was used to encode each part and you are responsible to figure out which element of this array corresponds to which part of the email and to decode these parts using the appropriate encoding. SendGrid's documentation is not very clear about this but it does mention it briefly:

> Messages and their headers can have character set data associated with them. In order to simplify the parsing of messages for the end user, SendGrid will decode the to, from, cc, and subject headers if needed. All headers will be converted to UTF-8 for uniformity since technically a header can be in many different character sets.
The charsets variable will contain a JSON encoded hash of the header / field name and its respective character set. For instance, it may look like:
[charsets] => {"to":"UTF-8","cc":"UTF-8","subject":"UTF-8","from":"UTF-8","text":"iso-8859-1"}
This shows that all headers should be treated as UTF-8, and the text body is latin1.

TL;DR; Instead of attempting to parse the webhook yourself, I suggest you use the `WebhookParser` which is built-in the [StrongGrid library](https://www.nuget.org/packages/StrongGrid/) (disclaimer: I'm the author of this library). This parser already handles the decoding of each part using the appropriate code page. See some [code examples](https://github.com/Jericho/StrongGrid#parser).

I hope this helps.

**Comment by thinkingserious** [comment]
Thank you @Jericho!

**Comment by ysunil702** [comment]
Thank you @Jericho
